include/deal.II/base/tensor_accessors.h

```diff
@@ -0,0 +1,586 @@
+   * invocations:
+   * @code
+   *   Tensor<5, dim> tensor;
+   *   tensor[0][1][2][3][4] = 42.;
+   * @endcode
+   * Index 1 (the 2nd index, count starts at 0) can now be shifted to the
+   * end via
+   * @code
+   *   auto tensor_view = reordered_index_view<1, 5>(tensor);
+   *   tensor_view[0][2][3][4][1] == 42.; // is true
+   * @endcode
+   * The usage of the dealii::Tensor type was solely for the sake of an
+   * example. The mechanism implemented by this function is available for
+   * fairly general tensorial types @tparam T.
+   *
+   * @note This function returns an internal class object consisting of an
+   * array subscript operator <code>operator[](unsigned int)</code> and a
+   * typedef <code>value_type</code> describing its return value.
+   *
+   * @tparam index The index to be shifted to the end. Indices are counted
+   * from 0, thus the valid range is $0\le\text{index}<\text{rank}$.
+   * @tparam rank Rank of the tensorial object @param t
+   * @tparam T A tensorial object of rank @tparam rank. @tparam T must
+   * provide a local typedef <code>value_type</code> and an index operator
+   * <code>operator[]()</code> that returns a (const or non-const)
+   * reference of <code>value_type</code>:
+   * @code
+   *   class T
+   *   {
+   *     typedef ... value_type
+   *     value_type & operator[](unsigned int);
+   *     const value_type & operator[](unsigned int) const;
+   *   };
+   * @endcode
+   *
+   * @relates ReorderedIndexView
+   */
+  template <int index, int rank, typename T>
+  internal::ReorderedIndexView<index, rank, T>
+  reordered_index_view(T &t)
+  {
+    return internal::ReorderedIndexView<index, rank, T>(t);
+  }
+
+
+  namespace internal
+  {
+    // Straightforward recursion implemented by specializing ExtractHelper
+    // for position == rank. We use the type trait ReturnType<rank, T> to
+    // have an idea what the final type will be.
+    template<int position, int rank>
+    struct ExtractHelper
+    {
+      template<typename T>
+      static
+      typename ReturnType<rank - position, T>::value_type &
+      extract(T &t, const TableIndices<rank> &indices)
+      {
+        return ExtractHelper<position + 1, rank>::template extract<typename AccessType<T>::value_type>(
+          t[indices[position]], indices);
+      }
+    };
+
+    // For dimension == rank there is nothing to extract, just return the
+    // object.
+    template<int rank>
+    struct ExtractHelper<rank, rank>
+    {
+      template<typename T>
+      static
+      T &
+      extract(T &t, const TableIndices<rank> &indices)
+      {
+        return t;
+      }
+    };
+  } /* namespace internal */
+
+
+  /**
+   * Return a reference (const or non-const) to a subobject of a tensorial
+   * object T as described by a TableIndices object @p indices. For
+   * example:
+   * @code
+   *   Tensor<5, dim> tensor;
+   *   TableIndices<5> indices (0, 1, 2, 3, 4);
+   *   TensorAccessors::extract(tensor, indices) = 42;
```

Comment:
As a matter of fact, you could even give it an `template <int lower_rank> operator[] (const TableIndices<lower_rank> &)` that returns a sub-tensor in case `lower_rank < rank`.